cassandra/connection.py

```diff
@@ -782,6 +785,20 @@ def _wrap_socket_from_context(self):
         self._socket = self.ssl_context.wrap_socket(self._socket, **ssl_options)
 
     def _initiate_connection(self, sockaddr):
+        if self.shard_id is not None:
+            # TODO: move to constructor
+            def port_generator():
+                for p in range(32768, 60999):
+                    if p % self.total_shards == self.shard_id:
+                        yield p
+
+            for port in port_generator():
+                try:
+                    self._socket.bind(('', port))
+                    break
+                except Exception as ex:
+                    print(ex)
```

Comment:
we need a `else` here in case we don't find a port right?

```suggestion
                    log.debug("Source port %s unavailable while connecting to host %s" % (port, self.endpoint))
            else:
                # TODO: decide what to do here
                raise Exception("Could not find a valid source local port while connecting to %s" % self.endpoint)
```

Reply:
no we don't, since if we don't get it we just don't bind. and it would fall on some shard. and the logic we have in place would close it and continue again.